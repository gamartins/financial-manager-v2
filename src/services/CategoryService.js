const key = 'fmv2:category'

const get = () => {
  const dataList = JSON.parse(localStorage.getItem(key))

  if (!dataList) {
    localStorage.setItem(key, JSON.stringify([]))
    return []
  }

  return dataList
}

const save = dataList => {
  localStorage.setItem(key, JSON.stringify(dataList))
}

export default {
  get,
  save,
}
