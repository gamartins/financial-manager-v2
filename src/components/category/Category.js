import React, { Component } from 'react'

import CategoryForm from './CategoryForm'
import CategoryList from './CategoryList'
import CategoryService from '../../services/CategoryService'

export default class Category extends Component {

  constructor(props) {
    super(props)

    this.state = {
      dataList: [],
      formValues: {
        id: null,
        name: '',
      }
    }
  }

  componentDidMount() {
    this.getAll()
  }

  getAll() {
    const dataList = CategoryService.get()
    this.setState({ dataList })
  }

  setFormValue(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index === -1) {
      return
    }

    const newFormValues = { ...this.state.dataList[index] }
    
    this.setState({ formValues: newFormValues })
  }

  resetFormValue() {
    this.setState({
      formValues: {
        id: null,
        name: '',
      }
    })
  }

  save(values) {
    return values.id ? this.edit(values) : this.add(values)
  }

  add(values) {
    const dataList = [ ...this.state.dataList ]
    dataList.push({
      id: Math.ceil((Math.random() * 100000)),
      name: values.name,
    })

    CategoryService.save(dataList)

    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  edit(values) {
    const index = this.state.dataList.findIndex(item => item.id === values.id)
    const dataList = [ ...this.state.dataList ]
    dataList[index] = {
      id: values.id,
      name: values.name
    }

    CategoryService.save(dataList)

    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  remove(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index !== -1) {
      this.state.dataList.splice(index, 1)
      this.setState({ dataList: this.state.dataList })
      this.resetFormValue()
    }    
  }

  render() {
    return (
      <div>
        <CategoryForm
          initialValues={this.state.formValues}
          saveItem={values => this.save(values)} />
        <CategoryList
          dataList={this.state.dataList}
          removeItem={id => this.remove(id)}
          editItem={item => this.setFormValue(item)} />
      </div>
    )
  }
}
