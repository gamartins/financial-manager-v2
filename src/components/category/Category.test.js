import React from 'react'
import Category from './Category'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

jest.mock('../../services/CategoryService');
import CategoryService from '../../services/CategoryService';

Enzyme.configure({ adapter: new Adapter() });

CategoryService.get.mockImplementation(() => [
    {
        id: 1,
        name: 'Category 01',
    },
    {
        id: 2,
        name: 'Category 02',
    },
])

describe('<Category />' ,() => {
    test('should be created', () => {
        const wrapper = mount(<Category />)
        expect(wrapper).toMatchSnapshot()
    })

    test('should get category list on componentDidMount', () => {
        const wrapper = shallow(<Category />)
        expect(wrapper.instance().state.dataList).toHaveLength(2)
    })

    test('should find and set form value with item selected', () => {
        const wrapper = shallow(<Category />)
        wrapper.instance().setFormValue(2)

        expect(wrapper.instance().state.formValues).toEqual({ id: 2, name: 'Category 02' })
    })

    test('should call add when category dont exists', () => {
        const wrapper = shallow(<Category />)
        const addMock = wrapper.instance().add = jest.fn()
        wrapper.instance().save({ id: null, name: 'Category 02' })

        expect(addMock).toHaveBeenCalled()
    })

    test('should call edit when category already exists', () => {
        const wrapper = shallow(<Category />)
        const editMock = wrapper.instance().edit = jest.fn()
        wrapper.instance().save({ id: 2, name: 'Category 02' })

        expect(editMock).toHaveBeenCalled()
    })

    test('should remove one element when call remove', () => {
        const wrapper = mount(<Category />)

        expect(wrapper.instance().state.dataList).toHaveLength(2)

        wrapper.instance().remove(2)

        expect(wrapper.instance().state.dataList).toHaveLength(1)
    })
})