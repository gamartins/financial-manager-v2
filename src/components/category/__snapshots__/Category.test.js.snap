// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Category /> should be created 1`] = `ReactWrapper {}`;
